pkg/ipam/ippool_test.go

```diff
@@ -0,0 +1,525 @@
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Globalnet IP Pool", func() {
+	Context("Pool Creation", testPoolCreation)
+	Context("IP Allocation", testIPAllocation)
+	Context("IP Release", testIPRelease)
+	Context("Block IPs", testBlockIPs)
+})
+
+const (
+	requestIP1    = "169.254.1.1"
+	service1      = "default/service1"
+	pod1          = "default/pod1"
+	pod2          = "default/pod2"
+	testCidr      = "169.254.1.1/30"
+	testBlockCidr = "169.254.1.1/28"
+	blockKey1     = "block1"
+	blockSize1    = 4
+	blockKey2     = "block2"
+	blockSize2    = 8
+)
+
+func testPoolCreation() {
+	When("CIDR is missing the prefix", func() {
+		It("should fail to create the IP pool with an error", func() {
+			pool, err := NewIPPool("169.254.0.0")
+			Expect(err).To(HaveOccurred())
+			Expect(pool).To(BeNil())
+		})
+	})
+	When("CIDR prefix is /33", func() {
+		It("should fail to create IP pool with an error", func() {
+			pool, err := NewIPPool("169.254.0.0/33")
+			Expect(err).To(HaveOccurred())
+			Expect(pool).To(BeNil())
+		})
+	})
+	When("CIDR Prefix is /32", func() {
+		It("should fail to create IP pool with an error", func() {
+			pool, err := NewIPPool("169.254.1.0/32")
+			Expect(err).To(HaveOccurred())
+			Expect(pool).To(BeNil())
+		})
+	})
+	When("CIDR Prefix is /31", func() {
+		It("should fail to create IP pool with an error", func() {
+			pool, err := NewIPPool("169.254.1.0/31")
+			Expect(err).To(HaveOccurred())
+			Expect(pool).To(BeNil())
+		})
+	})
+	When("CIDR Prefix is /30", func() {
+		It("should create IP pool of size 2", func() {
+			pool, err := NewIPPool("169.254.1.0/30")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(pool.available.Size()).Should(Equal(2))
+		})
+	})
+	When("CIDR Prefix is /24", func() {
+		It("should create IP pool of size 254", func() {
+			pool, err := NewIPPool("169.254.1.0/24")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(pool.available.Size()).Should(Equal(254))
+		})
+	})
+	When("CIDR Prefix is /16", func() {
+		It("should create an IP pool of size 65534", func() {
+			pool, err := NewIPPool("169.254.1.0/16")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(pool.available.Size()).Should(Equal(65534))
+		})
+	})
+	When("Any pool is created", func() {
+		It("should have allocated pool size of 0", func() {
+			pool, err := NewIPPool("169.254.1.0/16")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(pool.allocated)).Should(Equal(0))
+		})
+	})
+}
+
+func testIPAllocation() {
+	When("any IP is requested", testAnyIPRequested)
+	When("a specific IP is Requested", testSpecificIPRequested)
+	When("all IPs are allocated", testAllIpsAllocated)
+}
+
+func testAnyIPRequested() {
+	var (
+		pool        *IPPool
+		available   int
+		allocated   int
+		allocatedIP string
+	)
+
+	BeforeEach(func() {
+		var err error
+		pool, err = NewIPPool(testCidr)
+		Expect(err).NotTo(HaveOccurred())
+
+		available = pool.available.Size()
+		allocated = len(pool.allocated)
+
+		allocatedIP, err = pool.Allocate(service1)
```

Comment:
pool.Allocate() takes an integer as an argument, but we are passing a string in the unit tests?